Fixes #249 Add stocks route and controller

yinho999 opened the pull request:
Fixes #249 
Added
- seed stock data
- stock view
- stock controller
- controller testing for both view and json format

yinho999 commented:
Hello @pollygee. I've updated the stocks UI layout to better handle and display extensive stock data lists. Please let me know if this approach works well or we can revert back to the original design.

From
![image](https://github.com/user-attachments/assets/6282e22a-e59f-4ddf-9c53-277f69092d7c)

To
![image](https://github.com/user-attachments/assets/4c29c262-3a3f-4cdd-9a6a-1f864df96f31)

AbuMareBear commented on app/views/stocks/show.json.jbuilder:
```diff
@@ -0,0 +1 @@
+json.partial! "stocks/stock", stock: @stock
```
Can we remove this?

AbuMareBear commented on db/seeds.rb:
```diff
@@ -16,6 +16,27 @@
 
 school = School.find_or_create_by(name: "Test School")
 
+Stock.find_or_create_by(ticker: "AAPL",
```
This might be more readable if every key is on it's own line.

AbuMareBear commented on db/seeds.rb:
```diff
@@ -16,6 +16,27 @@
+  company_name: "Apple Inc.",
+  company_website: "https://www.apple.com",
+  description: "Apple Inc. specializes in the conceptualization, production, and distribution of smartphones," \
+  " personal computers, tablets, wearable technology, accessories, and various associated services. " \
```
Question: Do we need a long description like this? Might be more readable to have a shorter one line description if the requirements allow for it.

yinho999 replied:
I am not entirely sure about the structure of stocks. I can make it shorter if needed

pollygee replied:
For now this is just for development so a short description (does not have to be accurate) would work just fine.

AbuMareBear commented on test/controllers/stocks_controller_test.rb:
```diff
@@ -0,0 +1,40 @@
+require "test_helper"
+
+class StocksControllerTest < ActionDispatch::IntegrationTest
+  def setup
+    @stock = create(:stock)
+    @user = create(:user)
+    sign_in @user
+  end
```
It has been my experience that having test set up like this makes tests harder to write and maintain. In my opinion every test should have its own setup. DRY (Don't Repeat Yourself) doesn't apply to testing.

yinho999 commented on app/views/stocks/_stock.json.jbuilder:
```diff
@@ -0,0 +1,2 @@
+json.extract! stock, :id, :ticker, :stock_exchange, :company_name, :company_website, :created_at, :updated_at
```
I saw there is json version for school view as well, are we not going to use the json format?

AbuMareBear replied:
Yeah. The Rails generator creates them but If we don't need JSON responses it's best not to have to support them. I've removed some other JSON views in other PRs.

yinho999 commented on app/helpers/stocks_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module StocksHelper
```
There are other similar blank helpers in the same directory as well, should I remove those as well?

yinho999 commented:
@pollygee @AbuMareBear 

Thank you for reviewing my PR and giving such detailed responses. I have reverted the styles of the stock page and removed the JSON file and related tests, as we currently do not need them.

AbuMareBear commented on app/controllers/stocks_controller.rb:
```diff
@@ -0,0 +1,21 @@
+  before_action :set_stock, only: %i[show]
+  before_action :authenticate_user!
+
+  # GET /stocks or /stocks.json
```
Might be good to remove these comments.

AbuMareBear commented on test/controllers/stocks_controller_test.rb:
```diff
@@ -0,0 +1,19 @@
+require "test_helper"
+
+class StocksControllerTest < ActionDispatch::IntegrationTest
+  test "should get index" do
+    @stock = create(:stock)
+    @user = create(:user)
+    sign_in @user
+    get stocks_url
+    assert_response :success
+  end
```
Does using local variables work here? Also might be good to add spaces between the stages of the test for readability.

```
stock = create(:stock)
user = create(:user)
sign_in user

get stocks_url

assert_response :success
```

AbuMareBear commented on test/factories/stocks.rb:
```diff
@@ -1,4 +1,22 @@
 FactoryBot.define do
   factory :stock do
+    sequence(:ticker) { |n| "STOCK#{n}" }
+    stock_exchange { "NYSE" }
+    sequence(:company_name) { |n| "Sample Company#{n} Inc." }
+    company_website { "https://example.com" }
+    description { "A sample stock for testing" }
+    industry { "Technology" }
+    management { "CEO: John Doe" }
+    employees { 1000 }
+    sequence(:competitor_names) { |n| "Competitor #{n}, Competitor #{n + 1}" }
+    sales_growth { 5.5 }
+    industry_avg_sales_growth { 4.2 }
+    debt_to_equity { 0.3 }
+    industry_avg_debt_to_equity { 0.4 }
+    profit_margin { 15.2 }
+    industry_avg_profit_margin { 12.8 }
+    cash_flow { 1_000_000.00 }
+    debt { 500_000.00 }
+    price_cents { 12_500 } # $125.00
```
A best practice for factories is to only include required fields. A great task would be to remove any optional fields here and add validations in the stock model for attributes that are not optional. That could be a separate PR.
